lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,313 @@
+    -> KeyHash
+keyHashFromAccXPubIx accXPub ix =
+    hashVerificationKey $ deriveVerificationKey accXPub ix
+
+-- | The extension to the SharedState pool gap is done by adding next adjacent
+-- indices and their corresponding public keys, marking them as Unused.
+-- The number of added entries is determined by pool gap, and the start index is
+-- the next to the the specified index.
+extendSharedState
+    :: forall (n :: NetworkDiscriminant) k. Index 'Soft 'ScriptK
+    -> SharedState n k
+    -> SharedState n k
+extendSharedState ix state
+    | isOnEdge  = state { shareStateOurAddresses = nextAddresses }
+    | otherwise = state
+  where
+    edge = Map.size (shareStateOurAddresses state)
+    startIx = if edge == 0 then (toEnum 0) else succ ix
+    isOnEdge = edge - fromEnum ix <= fromEnum (shareStateGap state)
+    nextIndices =
+        L.take (fromEnum $ shareStateGap state) $ L.iterate succ startIx
+    (Right tag) = mkNetworkDiscriminant 1
+    nextAddresses
+      | ix == maxBound = mempty
+      | otherwise =
+            let pTemplate = shareStatePaymentTemplate state
+                dTemplate = shareStateDelegationTemplate state
+            in if templatesComplete pTemplate dTemplate then
+                foldr (addNotDiscoveredAddressToMap tag pTemplate dTemplate)
+                (shareStateOurAddresses state) nextIndices
+            else
+                mempty
+
+-- | The cosigner with his account public key is done per template.
+-- For every template the script is checked if the cosigner is present.
+-- If yes, then key is inserted. If the key is already present it is going to be
+-- updated. If there is no cosigner present is the script then the cosigner -
+-- account public key map is not changed.
+addCosignerAccXPub
+    :: forall (n :: NetworkDiscriminant) k. WalletKey k
+    => k 'AccountK XPub
+    -> Cosigner
+    -> SharedState n k
+    -> SharedState n k
+addCosignerAccXPub accXPub cosigner (SharedState ourAccXPub prefix g pT dT ourAdresesses) =
+    let updateScriptTemplate sc@(ScriptTemplate cosignerMap script') =
+            if cosigner `elem` retrieveAllCosigners script' then
+                ScriptTemplate (Map.insert cosigner (getRawKey accXPub) cosignerMap) script'
+            else
+                sc
+    in SharedState
+       { shareStateAccountKey = ourAccXPub
+       , shareStateDerivationPrefix = prefix
+       , shareStateGap = g
+       , shareStatePaymentTemplate = updateScriptTemplate pT
+       , shareStateDelegationTemplate = updateScriptTemplate <$> dT
+       , shareStateOurAddresses = ourAdresesses
+       }
+
+retrieveAllCosigners :: Script Cosigner -> [Cosigner]
+retrieveAllCosigners = foldScript (:) []
+
+isShared
+    :: (SoftDerivation k, WalletKey k)
+    => Address
+    -> SharedState n k
+    -> (Maybe (Index 'Soft 'ScriptK, KeyHash), SharedState n k)
+isShared address shared@(SharedState !accXPub !prefix !g !pT !dT !ourAddresses) =
+    case Map.lookup address ourAddresses of
```

Comment:
This won't work unfortunately :( ... For the reason stated in another comment. An address is ours if-and-only-if its payment part is ours. Please, let's strive to re-use code from `AddressPool`. 

We could do some pair-programming session to see how to abstract over the existing AddressPool if you want :+1:

Reply:
some attempt here https://github.com/input-output-hk/cardano-wallet/pull/2508/commits/a69d4db860f8f949f588b0533b06260c06a8966a